Replace Bindings map by dedicate `Bindings` interface. This allows to…

dmfs opened the pull request:
… compose the bindings from various sources without having to copy everything to a map. Implements #3

New bindings can adapt arbitrary sources, like Cursors, Android resources or Bundles, XML …

The following example show how the bindings can be composed from several sources:

```java
Bindings bindings = new Composite (
       new MapBindings(map1),
       new SingletonBindings("$json", new JsonObjectBindings(json)),
       new SingletonBindings("$map", new MapBindings(map2)),
   );
```

All the values in `map1` are accessible directly just like before

```
   {{ mapkey }}
```

The values in the JSON document and in `map2` are accessible with their prefixes

```
   {{ $json.key }}
   {{ $map.key }}
```

Also, in this commit:

In addition to the previous test, the following objects evaluate to true now:
* integers < 0
* non-empty CharSequences and Strings (the String check was broken before)
* non-empty Collections
* non-empty Maps and (nested) Bindings
* any other non-null and non-boolean value

Also, `for` blocks can now iterate arbitrary `Iterable`s.

Added a couple of test cases for the new code.

codeka commented on src/main/java/au/com/codeka/carrot/Bindings.java:
```diff
@@ -0,0 +1,13 @@
+ * @author marten
+ */
+public interface Bindings
+{
```
I know it might seem pointless, but I like to keep the code style the same throughout. Would you mind making sure all the { and on the previous line? "public interface Bindings {"

dmfs replied:
that's an absolutely reasonable request. In our projects we use a different code style and I'm just used to that.
If you use IntelliJ it might make sense to put your project code style settings under version control.

codeka replied:
Thanks for this. That's a great idea, I'll do that.

codeka commented on src/main/java/au/com/codeka/carrot/Scope.java:
```diff
@@ -36,12 +38,17 @@ public void pop() {
+   * @param bindings The new {@link Bindings}.
+   */
+  public void extendCurrent(Bindings bindings)
+  {
```
This one on the previous line, too. I won't bother commenting for the other places, but you get the idea :)

codeka commented on src/main/java/au/com/codeka/carrot/Bindings.java:
```diff
@@ -0,0 +1,13 @@
+ */
+public interface Bindings
+{
+    Object resolved(String key);
```
I think "resolve" (a verb) would be a better name.

Also, I would annotate this with @Nullable and document that null will be returned if the key doesn't exist.

dmfs replied:
sure, will do.

codeka commented on src/main/java/au/com/codeka/carrot/bindings/JsonArrayBindings.java:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @author Marten Gajda
+ */
+public final class JsonArrayBindings implements Bindings, Iterable
```
I'm thinking, given that we now have an easily-extensible interface that maybe this sort of specific sub-class could be kept out of the main library. Maybe we have a separate jar that implements this object for JSON data objects? I'm just thinking I'd like to keep the dependencies of the main library to a minimum (for example, if you're using Gson instead of org.json or something, then this would just be extra deps for no reason).

What do you think?

dmfs replied:
I was considering that. My idea was to give a brief example of how to create adapters for arbitrary sources.

The json.org dependency is added as [`compileOnly`](https://blog.gradle.org/introducing-compile-only-dependencies) which is the equivalent to Maven's `provided`. As a result, this library won't pull in `json.org`<sup>[1](#footnote)</sup> and you can not use this adapter unless you added the `json.org` dependency in your own project which uses carrot.

Does that make sense?

That being said I don't mind creating a sub-module if you prefer.


<a name="footnote">[1]</a>  see the pom which was created by jitpack for my branch: [carrot-stories~3-Bindings-v2.2.0-g2e98436-8.pom](https://jitpack.io/com/github/dmfs/carrot/stories~3-Bindings-v2.2.0-g2e98436-8/carrot-stories~3-Bindings-v2.2.0-g2e98436-8.pom), it contains `json.org` only as a test dependency

codeka replied:
Yeah, I think that's fine. Thanks for clarifying.

dmfs commented:
@codeka I've fixed the code style and added the annotations. Also I rebased on the new master and resolved the conflicts.

There is one more change in this version. Iterating over a `JSONObject` won't iterate over keys anymore like in the old version where you did this:
```
{% for key in json %}{{ key }} -> {{ json[key] }}
{% end %}
```
Instead it's iterating Map.Entries now so the template would look like this:
```
{% for item in json %}{{ item.key }} -> {{ item.value }}
{% end %}
```

I think that's closer to how you would do that in Jinja and Django.

I also updated the example in `README.md`.
